homeassistant/components/sleepiq/__init__.py

```diff
@@ -1,110 +1,227 @@
+
+    return unload_ok
 
-    def __init__(self, client):
-        """Initialize the data object."""
-        self._client = client
-        self.beds = {}
 
-        self.update()
+def _config_entry_update_signal_name(config_entry: ConfigEntry) -> str:
+    """Get signal name for updates to a config entry."""
+    return CONFIG_ENTRY_UPDATE_SIGNAL_TEMPLATE.format(config_entry.unique_id)
 
-    @Throttle(MIN_TIME_BETWEEN_UPDATES)
-    def update(self):
-        """Get the latest data from SleepIQ."""
-        self._client.login()
-        beds = self._client.beds_with_sleeper_status()
 
-        self.beds = {bed.bed_id: bed for bed in beds}
+async def _async_signal_options_update(
+    hass: HomeAssistant, config_entry: ConfigEntry
+) -> None:
+    """Signal config entry options update."""
+    async_dispatcher_send(
+        hass, _config_entry_update_signal_name(config_entry), config_entry
+    )
 
 
-class SleepIQSensor(Entity):
+class SleepIQDataUpdateCoordinator(DataUpdateCoordinator[dict[str, dict]]):
+    """SleepIQ data update coordinator."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        *,
+        client: Sleepyq,
+        update_interval: timedelta,
+        username: str,
+    ) -> None:
+        """Initialize coordinator."""
+        super().__init__(
+            hass, _LOGGER, name=f"{username}@SleepIQ", update_interval=update_interval
+        )
+        self.client = client
+
+    async def async_update_config(self, config_entry: ConfigEntry) -> None:
+        """Handle config update."""
+        self.update_interval = timedelta(
+            seconds=config_entry.options[CONF_SCAN_INTERVAL]
+        )
+
+    async def _async_update_data(self) -> dict[str, dict]:
+        data = {}
+        for bed in await self.hass.async_add_executor_job(
+            self.client.beds_with_sleeper_status
+        ):
+            data[bed.bed_id] = {BED: bed}
+
+        return data
+
+
+@dataclasses.dataclass
+class SleepIQHassData:
+    """Home Assistant SleepIQ runtime data."""
+
+    coordinators: dict[str, SleepIQDataUpdateCoordinator] = dataclasses.field(
+        default_factory=dict
+    )
+
+
+class SleepIQSensor(CoordinatorEntity):
     """Implementation of a SleepIQ sensor."""
 
-    def __init__(self, sleepiq_data, bed_id, side):
-        """Initialize the sensor."""
-        self._bed_id = bed_id
-        self._side = side
-        self.sleepiq_data = sleepiq_data
-        self.side = None
-        self.bed = None
+    def __init__(
+        self,
+        coordinator: SleepIQDataUpdateCoordinator,
+        bed_id: str,
+        side: str,
+    ) -> None:
+        """Initialize the SleepIQ side entity."""
+        super().__init__(coordinator)
+        self.bed_id = bed_id
+        self.side = side
 
-        # added by subclass
-        self._name = None
-        self.type = None
+        # Added by subclass.
+        self._name = ""
```

Comment:
```suggestion
        self._name: str | None = None
```

And assert where its used that its not None so you catch errors when its not been set